c/prima.c

```diff
@@ -44,19 +72,19 @@ int cobyla_c(const int m_nlcon, const prima_objcon_t calcfc, const void *data, c
              const int m_eq, const double Aeq[], const double beq[],
              const double xl[], const double xu[],
              double *f0, const double nlconstr0[],
-             int *nf, const double rhobeg, const double rhoend, const double ftarget, const int maxfun, const int iprint, const prima_callback_t callback, int *info);
+             int *nf, double *rhobeg, double *rhoend, const double ftarget, const int maxfun, const int iprint, const prima_callback_t callback, int *info);
```

Comment:
If a pointer is constant, I think it is better to say so explicitly, e.g., `double * const f0`.

Reply:
I find declaring a pointer const to be a little strange, since any modifications to the pointer within the function will be lost once the function returns. I think this looks a bit strange to most C programmers and doesn't provide any added benefit, so I think it's better to leave it as is. Although since the data is copied into a Fortran variable, I suppose I could put the original const back in along with the change to make it a pointer.

Reply:
> I find declaring a pointer const to be a little strange, since any modifications to the pointer within the function will be lost once the function returns. I think this looks a bit strange to most C programmers and doesn't provide any added benefit, so I think it's better to leave it as is.

My knowledge about C is probably outdated, but I do not agree with this point. We declare an argument to be a constant not because this declaration is useful in the sense that it empowers us to do something; instead, such a declaration is a safeguard that prevents the code from accidentally making unintentional changes to the argument (such code will not compile if we declare the argument as a constant); in addition, the `constant` keyword is also a kind of documentation, which promises to the user that the argument will never be changed by the function, and tells future contributors that they should never change the argument in the function. Sorry if my understanding is wrong.

Reply:
Two things:

1. The effect of `const` in `double * const f0` in the context of a function argument is not very sensible. It prevents users from doing something which I don't think any sensible user would ever try to do in the first place. I've made a short program with some comments to explain this further: https://godbolt.org/z/xa9aYh5j9

2. Since this is a declaration for a function which is defined in Fortran, declaring a const pointer doesn't have an effect in terms of compilation, since the code defining this function is in Fortran.

Actually, I thought about this quite a bit, and I think that there's a much simpler option for rhobeg/rhoend. They should be passed by value and Fortran should check for NaN, like this:

```fortran
real(C_DOUBLE), value, intent(IN) :: rhobeg
real(RP), allocatable :: rhobeg_loc

if(.not. isnan(rhobeg)) then
  rhobeg_loc = real(rhobeg, kind(rhobeg_loc))
end if
```

This is actually much better than passing these as a pointer, because passing as a pointer indicates both that the function may change the value, and that the user would have access to the changed value after the function call. However this is not the case since the contents of the pointer are copied. Passing by value indicates to the user that the program will do whatever it wants with the values without reporting it back to the user, which I think is appropriate here?

Reply:
> Two things:
> 
> 1. The effect of `const` in `double * const f0` in the context of a function argument is not very sensible. It prevents users from doing something which I don't think any sensible user would ever try to do in the first place. I've made a short program with some comments to explain this further: https://godbolt.org/z/xa9aYh5j9

>> such a declaration is a safeguard that prevents the code from accidentally making unintentional changes to the argument (such code will not compile if we declare the argument as a constant)

The keywords are "accidentally" and "unintentional" here. Most coding mistakes are made accidentally and unintentionally. Nobody is sensible when they make mistakes. 

> 2. Since this is a declaration for a function which is defined in Fortran, 

We have to think for the future. This is not necessarily always true in the future. 

> 
> Actually, I thought about this quite a bit, and I think that there's a much simpler option for rhobeg/rhoend. They should be passed by value and Fortran should check for NaN, like this:
> 
> ```fortran
> real(C_DOUBLE), value, intent(IN) :: rhobeg
> real(RP), allocatable :: rhobeg_loc
> 
> if(.not. isnan(rhobeg)) then
>   rhobeg_loc = real(rhobeg, kind(rhobeg_loc))
> end if
> ```
> 
> This is actually much better than passing these as a pointer, because passing as a pointer indicates both that the function may change the value, and that the user would have access to the changed value after the function call. However this is not the case since the contents of the pointer are copied. Passing by value indicates to the user that the program will do whatever it wants with the values without reporting it back to the user, which I think is appropriate here?

I understand the argument here. However, this involves changes to the Fortran source code, right? I am super reluctant to change the Fortran code unless there is a compelling reason. 

In addition, there is no way to have an `isnan` that works for all compilers under all options. See 

https://fortran-lang.discourse.group/t/challenge-testing-inf-and-nan-with-gfortran-13-ofast/6481
and
https://github.com/equipez/infnan

This is why `infnan.F90 and inf.F90` exists under `fortran/common` (PRIMA will unveil its complexity gradually when you know it more :) ). 

Indeed, I did exactly the same thing for `eta1,2` and `gamma1,2` before, but decided not to do it because `is_nan` may fail to work: 

https://github.com/libprima/prima/releases/tag/v0.7.2
 
This is also the motivation for the following test: 

https://github.com/primalib/prima/blob/main/.github/workflows/profile_infnan.yml#L104-L108

See particularly [L104--L108](https://github.com/primalib/prima/blob/main/.github/workflows/profile_infnan.yml#L104-L108).

Reply:
> I understand the argument here. However, this involves changes to the Fortran source code, right? I am super reluctant to change the Fortran code unless there is a compelling reason.

Only the `algorithm_c.f90` code. In the example I gave `rhobeg_loc` is allocatable, and if the if statement is false then it remains unallocated, and when it is passed to another subroutine `present` will return false.

It occurred to me that it might be possible to use `is_nan` instead of `present` for some values in the actual Fortran code, but I quickly decided to wouldn't be worth the trouble.

How worried are you about the issues with properly detecting NaN? And what would be the worst case scenario if an attempt to set rhobeg/end to NaN gets interpreted as a large value? In the interest of thinking for the future, maybe it's better to use pointers since null (to my knowledge) is a stable value, even if pointers can lead to non-intuitive function signatures.